### DISCUSSION TITLE
run .py file every reboot on server linux. its possible?

### DISCUSSION THREAD

**Discussion by Toei79** [GerryDazoo/Slinger#384] [General]
i try this but its not working, i have other server that runs fine , with one python3 too, idk whats wrong in this one. 

@reboot sudo python3 /home/ubuntu/Slinger/slingbox_server.py

**Comment by GerryDazoo** [comment]
You do not have to run it as root. post the server output. I suspect you're missing python modules

**Reply by Toei79** [reply]
sorry whats the server output?

in the command line runs fine
idk i just check some before , and root make safer to run , just my guess

**Reply by Toei79** [reply]
sudo apt-get install python3-dateutil

well i search in google and try this but not working.

i did this too but i get an empty file

@reboot /home/ubuntu/Slinger/slingbox_server.py >> /home/ubuntu/Slinger/slingbox_server_cron.log 2>&1

**Comment by tom76017** [comment]
My Debian works fine as a service.  I just set it up as documented in the README.

    sudo chmod +x /home/slingbox/slingbox_server.py
    sudo cp sling.service /etc/systemd/system/.
    # enable it
    sudo systemctl daemon-reload
    sudo systemctl enable sling.service
    sudo systemctl start sling.service

**Reply by Toei79** [reply]
i did that i run those commmands

yours works every reboot? without place the  .py in command line

im not too versed on this.  only way i know works for sure its using crontab without touching command line well thats the goal.

**Reply by DoogieHowitzerWMD** [reply]
Copying the sling.service file to /etc/systemd/system/ sets things up so that the server runs as a background process. With every subsequent boot of your system, your Slinger Server should automatically start in the background (not sure about whether it waits for a login, as I have the device running my server set to not require username and password entry).

If you have set that up correctly, you won't need to initiate the server to start by running it from the command line. However, if you still want to view the output code in real time, you can use `tail -f /tmp/sling.log`. In fact, you can use that command immediately after rebooting and verify that the server is running.

**Reply by Toei79** [reply]
> Copying the sling.service file to /etc/systemd/system/ sets things up so that the server runs as a background process. With every subsequent boot of your system, your Slinger Server should automatically start in the background (not sure about whether it waits for a login, as I have the device running my server set to not require username and password entry).
> 
> If you have set that up correctly, you won't need to initiate the server to start by running it from the command line. However, if you still want to view the output code in real time, you can use `tail -f /tmp/sling.log`. In fact, you can use that command immediately after rebooting and verify that the server is running.

Question , I only need add sling service to etc/systemd/system ? Sorry I'm not too versed, I know do that  for sure and check  log after reboot 

Thanks

**Reply by tom76017** [reply]
Don't forget to enable the sling.service once you copy it to etc/systemd/system.

Here's my sling.service.  I changed the config argument to test the E1: password, but the normal config.ini works just as good with the 15 char pw.

[Unit]
Description=SlingBox Server Service
After=multi-user.target

[Service]
Type=idle
User=slingbox
WorkingDirectory=/home/slingbox
StandardOutput=append:/tmp/sling.log
StandardError=append:/tmp/sling.log
ExecStart=/usr/bin/python3.9 -u slingbox_server.308f.py config-M1.0-LongPW.ini
Restart=always

[Install]
WantedBy=multi-user.target

**Reply by DoogieHowitzerWMD** [reply]
Yes, basically the stuff inside the sling.service file points to the locations of the files needed for running the server. The real, technical answer might be more complicated than that, but that is my simple understanding of it.

If you look at the text of the sling.service file, as long as what is defined as your "working directory" (line 10) is the location where you have your files for `sling_server.py`, your config file(s), and another identical copy of sling.service, then it should have everything it needs. The default is location of the working directory is `/home/slingbox` (what it should be if you followed the directions exactly as they are in the readme).

The only caveat is if you have more than one config file, or if your single config file is named anything other than exactly "config.ini", you will need to add it in the sling.service file on line 13.

So `ExecStart=/usr/bin/env python -u slingbox_server.py `

will need to look like `ExecStart=/usr/bin/env python -u slingbox_server.py your_config_file_name.ini`

**Reply by DoogieHowitzerWMD** [reply]
> Don't forget to enable the sling.service once you copy it to etc/systemd/system.
> 
> Here's my sling.service. I changed the config argument to test the E1: password, but the normal config.ini works just as good with the 15 char pw.
> 
> [Unit] Description=SlingBox Server Service After=multi-user.target
> 
> [Service] Type=idle User=slingbox WorkingDirectory=/home/slingbox StandardOutput=append:/tmp/sling.log StandardError=append:/tmp/sling.log ExecStart=/usr/bin/python3.9 -u slingbox_server.308f.py config-M1.0-LongPW.ini Restart=always
> 
> [Install] WantedBy=multi-user.target

I have `/usr/bin/env python...` for line 13 in my file.

I know that the way you have it is calling python3.9 directly, but how do I know what version of python is used with "/env python"?

The server works, and I presume that it is running on some version of python greater than 3.7 (the code in the server file itself looks like it has a line returning an error message if the version of python is below 3.7). I was just thinking about it, and I don't understand why it wouldn't need to be `/usr/bin/env python3...` in order to call some form of python3 and not a version of python 2 (which "python" by itself would make me think is implied).

**Reply by tom76017** [reply]
look at this to try to see what version your referencing.

**root@falcon:~$ which python**
/usr/bin/python

**root@falcon:~$ ls -al /usr/bin/python***
lrwxrwxrwx 1 root root       9 Apr  5  2021 /usr/bin/python -> python3.9
lrwxrwxrwx 1 root root       9 Jul 28  2021 /usr/bin/python2 -> python2.7
-rwxr-xr-x 1 root root 3602816 Jul 14  2021 /usr/bin/python2.7
lrwxrwxrwx 1 root root       9 Apr  5  2021 /usr/bin/python3 -> python3.9
-rwxr-xr-x 1 root root 5479736 Feb 28  2021 /usr/bin/python3.9
lrwxrwxrwx 1 root root      33 Feb 28  2021 /usr/bin/python3.9-config -> x86_64-linux-gnu-python3.9-config
lrwxrwxrwx 1 root root      16 Apr  5  2021 /usr/bin/python3-config -> python3.9-config

**Reply by GerryDazoo** [reply]
Doogie, you wrot:e  don't understand why it wouldn't need to be"/usr/bin/env python3.."
The short answer is it depends on how you installed python. If it was installed just for the slingbox user you'll see some python files in the user folder.  pyenv.cfg bin

**Reply by DoogieHowitzerWMD** [reply]
@GerryDazoo, Python wasn't something I added on Raspberry Pi, since it comes packaged with it on RPi OS. But from looking inside /usr/bin, what I was seeing listed was:

- python
- python3
- python3.9
- python3.9-config
- python3-config

so I wasn't sure how to determine what "python" was invoking in line 13 of the service file.

Then, by running `ls -al /usr/bin/python*`, the first two I see are `/usr/bin/python -> python3` and `/usr/bin/python3 -> python3.9`

Would it then follow that python -> python3 -> python3.9? I'm not sure if I am supposed to transitively interpret it that way.

**Reply by Toei79** [reply]
its not working for me. i able to move sling.service to that folder

then i let sling.service like this
[Unit]
Description=SlingBox Server Service
After=multi-user.target
Wants=systemd-hostnamed.service


[Service]
Type=idle
User=slingbox
WorkingDirectory=/home/ubuntu/Slinger
StandardOutput=append:/tmp/sling.log
StandardError=append:/tmp/sling.log
ExecStart=/usr/bin/python3 python3 -u slingbox_server.py 
Restart=always

[Install]
WantedBy=multi-user.target

i suppose i enable it before reboot to make run, otherwise its not necessary after reboot?

then


$ tail -f /tmp/sling.log
   tail: cannot open '/tmp/sling.log' for reading: No such file or directory
   tail: no files remaining

**Reply by Toei79** [reply]
> Copying the sling.service file to /etc/systemd/system/ sets things up so that the server runs as a background process. With every subsequent boot of your system, your Slinger Server should automatically start in the background (not sure about whether it waits for a login, as I have the device running my server set to not require username and password entry).
> 
> If you have set that up correctly, you won't need to initiate the server to start by running it from the command line. However, if you still want to view the output code in real time, you can use `tail -f /tmp/sling.log`. In fact, you can use that command immediately after rebooting and verify that the server is running.

oh always i required a login.  yea with password.

**Reply by GerryDazoo** [reply]
Just copying the file isn't sufficient. You have to tell systemd to use it.
Read the Linux notes in the readme file.

**Reply by tom76017** [reply]
The ExecStart line does not look right.

Your line
ExecStart=/usr/bin/python3 **python3** -u slingbox_server.py

Suggest you try:
ExecStart=/usr/bin/python3 -u slingbox_server.py

Then:
sudo cp sling.service /etc/systemd/system/.
sudo systemctl daemon-reload
sudo systemctl enable sling.service
sudo systemctl start sling.service

**Reply by Toei79** [reply]
> Just copying the file isn't sufficient. You have to tell systemd to use it. Read the Linux notes in the readme file.

i already enable it. i round all those commands from sling.service file

sudo systemctl enable /etc/systemd/system/sling.service


im going to try what tom suggest.  thanks

**Comment by Toei79** [comment]
![12-02-2022 (085327)](https://user-images.githubusercontent.com/79808610/205309648-1e64aea8-55c4-4521-aa0e-e3d515e8021f.jpg)

nothing its not working. i try play after reboot and nothings come.

also i check where python its and its there.

**Reply by tom76017** [reply]
Just try this from the command prompt.  Ctrl-C to stop it.   If that works, do the copy and reenable steps.

/usr/bin/python3 -u slingbox_server.py

**Reply by DoogieHowitzerWMD** [reply]
@Toei79 What is the name of the user you are using to run this on your computer?

If it isn't "slingbox", you need to change line 9 so that it reads `User="yourusername"`

**Reply by GerryDazoo** [reply]
Please post the output of the log file.
The service file was originally make for the Raspberry Pi distribution which is based on Debian. You may not have the same version of systemd running on Ubuntu.

**Reply by Toei79** [reply]
I don't where get a log i used the command without success

I will try what doogie says my ubuntu have other name

Ubuntupny

Thats my user ubuntu name

**Reply by GerryDazoo** [reply]
Please post the output from the command you ran.
the log file should be /tmp/sling.log

**Reply by Toei79** [reply]
Ok the one that I run works perfectly when I do in the command line

But this sling.service no yet I need change user name then see what happens 

Thanks for all the help

**Reply by tom76017** [reply]
Run this from the command prompt and post the output.

sudo systemctl status sling.service

**Reply by Toei79** [reply]
it was the name! amazing,  ubuntu user name instead slingbox in my case,  now its working so far.  thank you guys

i wish learn , whats the difference between use crontab and service? where use one or other one.  i google and says service uses more resources.
i have one microphone that i want run in this way but its not working with crontab maybe i give a try service.

thanks!

**Comment by socelillyhammer** [comment]
Hi.  Need some help with Pi.
Read and followed the README file.
I did not create a new user slingbox with new pw.  I am just keeping my current user name zyx
I created a subfolder for the slinger.py called slinger.  So I coped the sling.server; slngbox_server.py; and config.ini to this folder

/home/zyx/singer

Then I follow the readme to execute

sudo chmod +x /home/zyx/slinger/slingbox_server.py (changing the folder directory)

but when I enter this ---- it cannot execute and I cannot move forward

sudo cp sling.service /etc/systemd/system/.


Any help here?

I can run the .py file itself and I can work and see and hear audio.  but I cannot finish following the readme.

This is where I am at.  Thank you in advance for the help

* Copy "slingbox_server.py, config.ini and sling.service to /home/slingbox   (changed it to /home/zyx/slinger) --- i also altered the sling.server file WorkingDirectory to read /home/zyx/slinger

```sh
    sudo chmod +x /home/slingbox/slingbox_server.py
    sudo cp sling.service /etc/systemd/system/.
    # enable it
    sudo systemctl daemon-reload
    sudo systemctl enable sling.service
    sudo systemctl start sling.service
    
    sudo systemctl stop sling.service # to shut it down   
```

* The default config generates a log file /tmp/sling.log   to check to see what's going on....

As root, add or modify the two following lines to your /etc/sysctl.conf 

```py
    net.core.rmem_max = 8192000
    net.core.wmem_max = 8192000

**Reply by GerryDazoo** [reply]
Next time can you post the actually output from the Pi so we can see EXACTLY what the error is?
That said, the command to make the .py file executable is not needed. 
sudo chmod +x /home/slingbox/slingbox_server.py
I'll fix the readme. It's something that was added by someone e;se and I didn't notice. I'll fix the readme
And yes you will need to edit the service file to include the correct path to your slingbox_server.py file.

Also, I now see the exact error your getting when trying to use the chmod command. It's because you are using the wrong path to the server file. You neglected to include the "xyz".

**Reply by socelillyhammer** [reply]
Thanks.  I've been executing it via remote ssh by typing ./slingbox_server.py when i am in the home directory. the sudo chmod command doesnt run it for some odd reason..  But, whatever works, right?

**Comment by pfuller88** [comment]
my slinger.service file has an env in the , try this....


ExecStart=/usr/bin/env python3 -u slingbox_server.py

**Reply by socelillyhammer** [reply]
Thank you.  Figured it out.  I had to mess around with the user name and path in the sling.server file.  I think this is what @GerryDazoo was referencing to me earlier that i can mess with this file to have the server start automatically upon booting up the Pi and then by clicking "restart" on the remote it will restart the server.

**Reply by tom76017** [reply]
This is is also an error in setup.   You need to be in the directory with the sling.service file when the cp is executed.
![image](https://user-images.githubusercontent.com/26171888/209140479-c26826a8-9989-43d0-9a52-948e3d4789b1.png)
